### DISCUSSION TITLE
Board field detection

### DISCUSSION THREAD

**Discussion by wormstein** [EdNekebno/DGTCentaurMods#13] [General]
Hi guys,
In the quest to find more events of the board I have one that I cannot reverse engineer.

I cannot figure it out how the board reads the status of the fields  when the centaur software starts. Have you tried to catch this event?

To trigger it place the pieces randomly on the board and start centaur software. It will highlight fields on the board where pieces are not in the correct place for a new game. It's clear that the board gets back the board state from the controller just after start.

Sniffing serial I see a command after software starts and I suspect the respose contains that data. I cannot figure it out how it is  constructed... I think for each field board reports x02 for correct state and x03 for wrong so board will highlight those fields. Correct state meaning pieces should be in start position,

Have a look yourself here, maybe you have other ideas...

662   write(7</dev/ttyS0>, "\xf4\x00\x07\x06\x50\x7f\x50", 7) = 7
662   read(7</dev/ttyS0>, "\xf4", 1)    = 1
662   read(7</dev/ttyS0>, "\x01", 1)    = 1
662   read(7</dev/ttyS0>, "\x07", 1)    = 1
662   read(7</dev/ttyS0>, "\x06", 1)    = 1
662   read(7</dev/ttyS0>, "P", 1)       = 1
662   read(7</dev/ttyS0>, "\x7f", 1)    = 1
662   read(7</dev/ttyS0>, "\x02", 1)    = 1
662   read(7</dev/ttyS0>, "=", 1)       = 1
662   read(7</dev/ttyS0>, "\x02", 1)    = 1
662   read(7</dev/ttyS0>, "9", 1)       = 1
662   read(7</dev/ttyS0>, "\x02", 1)    = 1
662   read(7</dev/ttyS0>, ";", 1)       = 1
662   read(7</dev/ttyS0>, "\x02", 1)    = 1
662   read(7</dev/ttyS0>, ">", 1)       = 1
662   read(7</dev/ttyS0>, "\x02", 1)    = 1
662   read(7</dev/ttyS0>, "=", 1)       = 1
662   read(7</dev/ttyS0>, "\x02", 1)    = 1
662   read(7</dev/ttyS0>, "A", 1)       = 1
662   read(7</dev/ttyS0>, "\x02", 1)    = 1
662   read(7</dev/ttyS0>, "A", 1)       = 1
662   read(7</dev/ttyS0>, "\x02", 1)    = 1
662   read(7</dev/ttyS0>, "x", 1)       = 1
662   read(7</dev/ttyS0>, "\x02", 1)    = 1
662   read(7</dev/ttyS0>, "4", 1)       = 1
662   read(7</dev/ttyS0>, "\x02", 1)    = 1
662   read(7</dev/ttyS0>, "0", 1)       = 1
662   read(7</dev/ttyS0>, "\x02", 1)    = 1
662   read(7</dev/ttyS0>, "9", 1)       = 1
662   read(7</dev/ttyS0>, "\x02", 1)    = 1
662   read(7</dev/ttyS0>, "8", 1)       = 1
662   read(7</dev/ttyS0>, "\x02", 1)    = 1
662   read(7</dev/ttyS0>, "6", 1)       = 1
662   read(7</dev/ttyS0>, "\x02", 1)    = 1
662   read(7</dev/ttyS0>, "=", 1)       = 1
662   read(7</dev/ttyS0>, "\x02", 1)    = 1
662   read(7</dev/ttyS0>, "@", 1)       = 1
662   read(7</dev/ttyS0>, "\x03", 1)    = 1
662   read(7</dev/ttyS0>, "9", 1)       = 1
662   read(7</dev/ttyS0>, "\x02", 1)    = 1
662   read(7</dev/ttyS0>, "6", 1)       = 1
662   read(7</dev/ttyS0>, "\x02", 1)    = 1
662   read(7</dev/ttyS0>, ".", 1)       = 1
662   read(7</dev/ttyS0>, "\x02", 1)    = 1
662   read(7</dev/ttyS0>, "7", 1)       = 1
662   read(7</dev/ttyS0>, "\x02", 1)    = 1
662   read(7</dev/ttyS0>, "8", 1)       = 1
662   read(7</dev/ttyS0>, "\x02", 1)    = 1
662   read(7</dev/ttyS0>, "9", 1)       = 1
662   read(7</dev/ttyS0>, "\x02", 1)    = 1
662   read(7</dev/ttyS0>, ">", 1)       = 1
662   read(7</dev/ttyS0>, "\x02", 1)    = 1
662   read(7</dev/ttyS0>, "I", 1)       = 1
662   read(7</dev/ttyS0>, "\x03", 1)    = 1
662   read(7</dev/ttyS0>, "X", 1)       = 1
662   read(7</dev/ttyS0>, "\x02", 1)    = 1
662   read(7</dev/ttyS0>, "1", 1)       = 1
662   read(7</dev/ttyS0>, "\x02", 1)    = 1
662   read(7</dev/ttyS0>, "/", 1)       = 1
662   read(7</dev/ttyS0>, "\x02", 1)    = 1
662   read(7</dev/ttyS0>, "3", 1)       = 1
662   read(7</dev/ttyS0>, "\x02", 1)    = 1
662   read(7</dev/ttyS0>, "7", 1)       = 1
662   read(7</dev/ttyS0>, "\x02", 1)    = 1
662   read(7</dev/ttyS0>, "6", 1)       = 1
662   read(7</dev/ttyS0>, "\x02", 1)    = 1
662   read(7</dev/ttyS0>, "=", 1)       = 1
662   read(7</dev/ttyS0>, "\x03", 1)    = 1
662   read(7</dev/ttyS0>, "5", 1)       = 1
662   read(7</dev/ttyS0>, "\x02", 1)    = 1
662   read(7</dev/ttyS0>, "g", 1)       = 1
662   read(7</dev/ttyS0>, "\x02", 1)    = 1
662   read(7</dev/ttyS0>, ",", 1)       = 1
662   read(7</dev/ttyS0>, "\x02", 1)    = 1
662   read(7</dev/ttyS0>, "/", 1)       = 1
662   read(7</dev/ttyS0>, "\x02", 1)    = 1
662   read(7</dev/ttyS0>, "/", 1)       = 1
662   read(7</dev/ttyS0>, "\x02", 1)    = 1
662   read(7</dev/ttyS0>, "3", 1)       = 1
662   read(7</dev/ttyS0>, "\x02", 1)    = 1
662   read(7</dev/ttyS0>, "8", 1)       = 1
662   read(7</dev/ttyS0>, "\x02", 1)    = 1
662   read(7</dev/ttyS0>, ";", 1)       = 1
662   read(7</dev/ttyS0>, "\x02", 1)    = 1
662   read(7</dev/ttyS0>, "8", 1)       = 1
662   read(7</dev/ttyS0>, "\x02", 1)    = 1
662   read(7</dev/ttyS0>, "F", 1)       = 1
662   read(7</dev/ttyS0>, "\x02", 1)    = 1
662   read(7</dev/ttyS0>, "0", 1)       = 1
662   read(7</dev/ttyS0>, "\x02", 1)    = 1
662   read(7</dev/ttyS0>, "5", 1)       = 1
662   read(7</dev/ttyS0>, "\x02", 1)    = 1
662   read(7</dev/ttyS0>, "3", 1)       = 1
662   read(7</dev/ttyS0>, "\x02", 1)    = 1
662   read(7</dev/ttyS0>, "7", 1)       = 1
662   read(7</dev/ttyS0>, "\x02", 1)    = 1
662   read(7</dev/ttyS0>, "4", 1)       = 1
662   read(7</dev/ttyS0>, "\x02", 1)    = 1
662   read(7</dev/ttyS0>, "9", 1)       = 1
662   read(7</dev/ttyS0>, "\x02", 1)    = 1
662   read(7</dev/ttyS0>, "=", 1)       = 1
662   read(7</dev/ttyS0>, "\x02", 1)    = 1
662   read(7</dev/ttyS0>, "M", 1)       = 1
662   read(7</dev/ttyS0>, "\x02", 1)    = 1
662   read(7</dev/ttyS0>, "+", 1)       = 1
662   read(7</dev/ttyS0>, "\x02", 1)    = 1
662   read(7</dev/ttyS0>, "'", 1)       = 1
662   read(7</dev/ttyS0>, "\x02", 1)    = 1
662   read(7</dev/ttyS0>, "C", 1)       = 1
662   read(7</dev/ttyS0>, "\x02", 1)    = 1
662   read(7</dev/ttyS0>, "6", 1)       = 1
662   read(7</dev/ttyS0>, "\x02", 1)    = 1
662   read(7</dev/ttyS0>, "B", 1)       = 1
662   read(7</dev/ttyS0>, "\x02", 1)    = 1
662   read(7</dev/ttyS0>, "@", 1)       = 1
662   read(7</dev/ttyS0>, "\x02", 1)    = 1
662   read(7</dev/ttyS0>, "w", 1)       = 1
662   read(7</dev/ttyS0>, "\x02", 1)    = 1
662   read(7</dev/ttyS0>, "E", 1)       = 1
662   read(7</dev/ttyS0>, "\x02", 1)    = 1
662   read(7</dev/ttyS0>, "U", 1)       = 1
662   read(7</dev/ttyS0>, "\x03", 1)    = 1
662   read(7</dev/ttyS0>, "\x15", 1)    = 1
662   read(7</dev/ttyS0>, "\x02", 1)    = 1
662   read(7</dev/ttyS0>, "e", 1)       = 1
662   read(7</dev/ttyS0>, "\x02", 1)    = 1
662   read(7</dev/ttyS0>, "G", 1)       = 1
662   read(7</dev/ttyS0>, "\x02", 1)    = 1
662   read(7</dev/ttyS0>, "?", 1)       = 1
662   read(7</dev/ttyS0>, "\x02", 1)    = 1
662   read(7</dev/ttyS0>, "K", 1)       = 1
662   read(7</dev/ttyS0>, "\x03", 1)    = 1
662   read(7</dev/ttyS0>, "\x7f", 1)    = 1
662   read(7</dev/ttyS0>, "\x02", 1)    = 1
662   read(7</dev/ttyS0>, "\x7f", 1)    = 1
662   read(7</dev/ttyS0>, "\x1c", 1)    = 1

**Comment by EdNekebno** [comment]
Yup. I see the feature if I start the centaur process. If I write that read command to the board in code though then I get the same response regardless of where the pieces are.

It seems to be a mode of the board. __main__.centaur.check_start_position() puts it into this mode that does the flashing lights thing. Interestingly if you call it you can then move pieces to the wrong places and their lights start flashing too. So I don't think the data gets sent back.

**Reply by EdNekebno** [reply]
Actually..ignore check_start_position() - I can't get it to do anything other than talk to stockfish.

**Comment by EdNekebno** [comment]
On a further look...on startup we have the following writes:

rite(7<\x2f\x64\x65\x76\x2f\x74\x74\x79\x53\x30>, "\x4d", 1) = 1
write(7<\x2f\x64\x65\x76\x2f\x74\x74\x79\x53\x30>, "\x4e", 1) = 1
write(7<\x2f\x64\x65\x76\x2f\x74\x74\x79\x53\x30>, "\xf4\x00\x07\x06\x50\x7f\x50", 7) = 7 - calibration data?
write(7<\x2f\x64\x65\x76\x2f\x74\x74\x79\x53\x30>, "\xf0\x00\x07\x06\x50\x7f\x4c", 7) = 7 - board data?
write(7<\x2f\x64\x65\x76\x2f\x74\x74\x79\x53\x30>, "\xb0\x00\x07\x06\x50\x00\x0d", 7) = 7 - this is ledsOff()
write(7<\x2f\x64\x65\x76\x2f\x74\x74\x79\x53\x30>, "\xb0\x00\x07\x06\x50\x00\x0d", 7) = 7 - ledsOff()
write(7<\x2f\x64\x65\x76\x2f\x74\x74\x79\x53\x30>, "\xb1\x00\x08\x06\x50\x4c\x08\x63", 8) = 8 - beep
write(7<\x2f\x64\x65\x76\x2f\x74\x74\x79\x53\x30>, "\xb0\x00\x0e\x06\x50\x05\x0a\x00\x05\x15\x17\x0d\x0f\x70", 14) = 14 - all light commands start \xb0
write(7<\x2f\x64\x65\x76\x2f\x74\x74\x79\x53\x30>, "\xb0\x00\x0b\x06\x50\x05\x02\x00\x05\x3f\x5c", 11) = 11

Then I think you're right but the important data is the response to the second request. If I dump the responses to the one I marked "calibration data?" and "board data?" with 0 pieces I get:
f4010706507f020e0206020a0205020a020f0212023b02130210021602120214021402120300021102100215021702160215021b031c02150210021402130214021c0308023b020b020f0211021402190219021a02250212020f020e02170217021b021d0220021302110226021a02200222024e02240237027902480227022802300358025c79

and

f0010706507f7f747f7f7f717f717f717f7c7f6f7f7f7f657f6f00017f797f757f7d7f767f7d7f727f6c7f7300007f7600037f757f717f617f797f7a7f717f6d7f777f737f707f6d7f707f7b7f6c7f7b7f737f6d7f797f6a7f797f787f737f707f6d7f707f707f757f717f787f707f7600007f737f6f7f777f7b7f7e7f737f767f747f777f7c39

The second one certainly has some repetition. These seem to match up with the function calls:

__main__.serial.send_calval() matches the one in your post and the one I marked "calibration data?" and __main__.serial.send_strength() matches the other one. Then it could be that the data is actually the strength of the field for each of the sensors on the board? I haven't yet worked out the format though if that's the case.

**Comment by wormstein** [comment]
Ahhh, I just noticed that github is horrible in formatting my post. I just edited my initial post...

The answer could be in any of the responses,,, at least in the first one there are 65 repetitions of x02 and x03 in total, so looks like the board responds something for each field.

Offtopic: what do you use to dump communication?

**Comment by EdNekebno** [comment]
Yup...there's patterns. In the second one:

f0010706507f
7f74
7f7f
7f71
7f71
7f71
7f7c
7f6f
7f7f
7f65
7f6f
0001
7f79
7f75
7f7d
7f76
7f7d
7f72
7f6c
7f73
0000
7f76
0003
7f75
7f71
7f61
7f79
7f7a
7f71
7f6d
7f77
7f73
7f70
7f6d
7f70
7f7b
7f6c
7f7b
7f73
7f6d
7f79
7f6a
7f79
7f78
7f73
7f70
7f6d
7f70
7f70
7f75
7f71
7f78
7f70
7f76
0000
7f73
7f6f
7f77
7f7b
7f7e
7f73
7f76
7f74
7f77
7f7c
39

First row is the request repeated with a 0x01 in the second position and no repeat of the checksum. 64 x 2 bytes. Then a checksum. I think the first one takes the same pattern but haven't checked it but that would fit your theory.

I use strace to dump the codes. Then I coded in python to send the request to the board so I can easily test what comes back - in this case I just hex dumped it this way because it's easier to see patterns.

**Comment by EdNekebno** [comment]
The data is in the second one....with four rooks on the board I get:

f0010706507f
055e7f697f637f7f7f757f687f750571
7f727f6c7f767f6e7f787f7b7f767f7c
7f697f7d7f7d7f797f797f747f7f7f6e
7f747f767f6f00017f6a7f717f707f77
7f777f777f7d7f707f757f7000017f69
7f7700057f6a7f777f777f7d7f787f7e
7f657f697f797f717f797f777f6f7f75
063f7f787f7c7f6e7f707f757f77063b
44850006065061

Notice the 2 bytes of each corner (055e, 0571, 063f, 063b) are all substantially lower than the 7fxx of the others. The exception being a couple of numbers are very low - I think this is board accuracy and probably what the calibration data is for.  These are always very low and change if I read again - for example on re-reading I get:

f0010706507f
05567f757f757f737f747f727f720571
7f727f717f6e7f747f7a7f7f7f727f7c
7f6e7f6d7f7b7f767f657f727f6e7f6b
7f777f6e7f757f787f6c7f6b7f7b7f70
7f6a7f7d7f717f6e7f637f667f6d7f68
7f727f6f7f757f747f7b7f7e7f757f7a
7f7a7f727f797f777f787f717f6d7f6b
06567f767f797f6300017f7b7f79063f
6fb1000606500d

**Comment by wormstein** [comment]
All these bytes gives a serious headache :))
I say you're right on it. The second one is the one that we have to dig...

I will be playing more on this. I just need to improve some of my tools here :)

In the meantime I've reversed some other calls, things like led intensity and how the board signals when starting four rows are completed. I will put these in a new discussion,

-Adrrian

**Comment by EdNekebno** [comment]
Sounds great.
I pushed a little something experimental for you to try: getBoardState() and printBoardState() . I completely ignored the calibration and just put variables in for upperlimit and lowerlimit - so they may need tweaking if my board is different to others.

**Comment by wormstein** [comment]
Nice done. Will play with it and see.

It's clear that the board electronics know nothing about chess other than reporting field state. All the logic is in the software...
